pallets/delegation/src/lib.rs

```diff
@@ -481,6 +538,72 @@ pub mod pallet {
 			)
 			.into())
 		}
+
+		/// Remove a delegation node (potentially a root node) and all its
+		/// children.
+		///
+		/// Removing a delegation node results in the trust hierarchy starting
+		/// from the given node being removed. Nevertheless, removal starts
+		/// from the leave nodes upwards, so if the operation ends prematurely
+		/// because it runs out of gas, the delegation state would be consistent
+		/// as no child would "survive" its parent. As a consequence, if the
+		/// given node is removed, the trust hierarchy with the node as root is
+		/// to be considered removed.
+		///
+		/// The dispatch origin must be `DelegationEntityId`.
+		///
+		/// Emits C * `DelegationRevoked`.
+		///
+		/// # <weight>
+		/// Weight: O(C) where C is the number of children of the delegation
+		/// node which is bounded by `max_children`.
+		/// - Reads: [Origin Account], Roots, C * Delegations, C * Children.
+		/// - Writes: Roots, 2 * C * Delegations
+		/// # </weight>
+		#[pallet::weight(
+			<T as Config>::WeightInfo::remove_delegation_root_child(*max_removals)
+				.max(<T as Config>::WeightInfo::remove_delegation_leaf(*max_removals)))]
+		pub fn remove_delegation(
+			origin: OriginFor<T>,
+			delegation_id: DelegationNodeIdOf<T>,
+			max_removals: u32,
+		) -> DispatchResultWithPostInfo {
+			let source = <T as Config>::EnsureOrigin::ensure_origin(origin)?;
+			let sender = source.sender();
+			let invoker = source.subject();
+
+			let delegation = <DelegationNodes<T>>::get(&delegation_id).ok_or(Error::<T>::DelegationNotFound)?;
+			let deposit_owner = delegation.deposit.owner;
+
+			// Node can only be removed by deposit or node owner, not the parent or another
+			// ancestor
+			ensure!(
+				deposit_owner == sender || deposit_owner == invoker.clone().into(),
+				Error::<T>::UnauthorizedRemoval
```

Comment:
DID will never own the deposit. For the invoker we should check `is_delegating`?

```suggestion
				deposit_owner == sender,
				Error::<T>::UnauthorizedRemoval
```

Reply:
I think we are only allowing the direct owner of a node, so probably not `is_delegating` but a simple check for the delegation owner == DID.